flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamSource.java

```diff
@@ -106,8 +106,10 @@ public void run(final Object lockingObject,
 			if (!isCanceledOrStopped()) {
 				advanceToEndOfEventTime();
 
+				// in theory, the subclasses of StreamSource may implement the BoundedOneInput interface,
+				// so we still need the following call to end the input
 				synchronized (lockingObject) {
-					operatorChain.endInput(1);
+					operatorChain.endOperatorInput(this, 1);
```

Comment:
`streamOperator` as head operator could be got directly from `operatorChain`, so we could provide `operatorChain#endHeadOperatorInput(int inputIndex)` to avoid passing `streamOperator` in the constructor.